code/datums/components/riding.dm

```diff
@@ -0,0 +1,441 @@
+
+/datum/component/riding/proc/handle_vehicle_layer()
+	if(_parent.dir != NORTH)
+		_parent.layer = ABOVE_MOB_LAYER
+	else
+		_parent.layer = OBJ_LAYER
+
+/datum/component/riding/proc/vehicle_moved()
+	for(var/mob/living/M in _parent.buckled_mobs)
+		ride_check(M)
+	handle_vehicle_offsets()
+	handle_vehicle_layer()
+
+/datum/component/riding/proc/ride_check(mob/living/M)
+	return TRUE
+
+/datum/component/riding/proc/force_dismount(mob/living/M)
+	_parent.unbuckle_mob(M)
+
+/datum/component/riding/proc/handle_vehicle_offsets()
+	var/_parent_dir = "[_parent.dir]"
+	var/passindex = 0
+	if(_parent.has_buckled_mobs())
+		for(var/m in _parent.buckled_mobs)
+			passindex++
+			var/mob/living/buckled_mob = m
+			var/list/offsets = get_offsets(passindex)
+			var/rider_dir = get_rider_dir(passindex)
+			buckled_mob.setDir(rider_dir)
+			dir_loop:
+				for(var/offsetdir in offsets)
+					if(offsetdir == _parent_dir)
+						var/list/diroffsets = offsets[offsetdir]
+						buckled_mob.pixel_x = diroffsets[1]
+						if(diroffsets.len >= 2)
+							buckled_mob.pixel_y = diroffsets[2]
+						if(diroffsets.len == 3)
+							buckled_mob.layer = diroffsets[3]
+						break dir_loop
+
+//Override this to set your vehicle's various pixel offsets
+/datum/component/riding/proc/get_offsets(pass_index) // list(dir = x, y, layer)
+	return list("[NORTH]" = list(0, 0), "[SOUTH]" = list(0, 0), "[EAST]" = list(0, 0), "[WEST]" = list(0, 0))
+
+//Override this to set the passengers/riders dir based on which passenger they are.
+//ie: rider facing the vehicle's dir, but passenger 2 facing backwards, etc.
+/datum/component/riding/proc/get_rider_dir(pass_index)
+	return _parent.dir
+
+//KEYS
+/datum/component/riding/proc/keycheck(mob/user)
+	if(keytype)
+		if(user.is_holding_item_of_type(keytype))
+			return TRUE
+	else
+		return TRUE
+	return FALSE
+
+//BUCKLE HOOKS
+/datum/component/riding/proc/restore_position(mob/living/buckled_mob)
+	if(istype(buckled_mob))
+		buckled_mob.pixel_x = 0
+		buckled_mob.pixel_y = 0
+		if(buckled_mob.client)
+			buckled_mob.client.change_view(world.view)
+
+//MOVEMENT
+/datum/component/riding/proc/handle_ride(mob/user, direction)
+	if(user.incapacitated())
+		Unbuckle(user)
+		return
+
+	if(world.time < next_vehicle_move)
+		return
+	next_vehicle_move = world.time + vehicle_move_delay
+	if(keycheck(user))
+		if(!Process_Spacemove(direction) || !isturf(_parent.loc))
+			return
+		step(_parent, direction)
+
+		handle_vehicle_layer()
+		handle_vehicle_offsets()
+	else
+		to_chat(user, "<span class='notice'>You'll need the keys in one of your hands to drive \the [_parent.name].</span>")
+
+/datum/component/riding/proc/Unbuckle(atom/movable/M)
+	addtimer(CALLBACK(_parent, /atom/movable/.proc/unbuckle_mob, M), 0, TIMER_UNIQUE)
+
+/datum/component/riding/proc/Process_Spacemove(direction)
+	if(_parent.has_gravity())
+		return 1
+
+	return 0
+
+/datum/component/riding/space/Process_Spacemove(direction)
+	return 1
+
+
+//atv
+/datum/component/riding/atv

```

Comment:
Instead of subtypes, you should just change the base type's vars based on Initialize parameters